crates/typst-pdf/src/lib.rs

```diff
@@ -61,383 +74,482 @@ pub fn pdf(
+///
+/// In other words: this struct follows the **typestate pattern**. This prevents
+/// you from using data that is not yet available, at the type level.
+///
+/// Each phase consists of processes, that can read the state of the previous
+/// phases (`S`), and write to the new state (`B`). These processes are run
+/// in the order [`PdfBuilder::run`] is called on them.
+///
+/// To combine the last state and what was built by the previous phases in a new
+/// state, [`PdfBuilder::transition`] is used. A new phase is then entered.
+///
+/// A final step, that has direct access to the global reference allocator and
+/// PDF document, can be run with [`PdfBuilder::export_with`].
+struct PdfBuilder<S, B> {
+    /// The context that has been accumulated so far.
+    state: S,
+    /// A new part of the context that is currently being built.
+    building: B,
+    /// A global bump allocator.
+    alloc: Ref,
+    /// The PDF document that is being written.
     pdf: Pdf,
+}
+
+/// The initial state: we are exploring the document, collecting all resources
+/// that will be necessary later.
+///
+/// The only step here is [`Pages`].
+struct BuildContent<'a> {
+    document: &'a Document,
+}
+
+/// All the resources that have been collected when traversing the document.
+///
+/// This does not allocate references to resources, only track what was used
+/// and deduplicate what can be deduplicated.
+///
+/// You may notice that this structure is a tree: [`PatternRemapper`] and
+/// [`ColorFontMap`] (that are present in the fields of [`Resources`]),
+/// themselves contain [`Resources`] (that will be called "sub-resources" from
+/// now on). Because color glyphs and patterns are defined using content
+/// streams, just like pages, they can refer to resources too, which are tracked
+/// by the respective sub-resources.
+///
+/// Each instance of this structure will become a `/Resources` dictionnary in
+/// the final PDF. It is not possible to use a single shared dictionnary for all
+/// pages, patterns and color fonts, because if a resource is listed in its own
+/// `/Resources` dictionnary, some PDF readers will fail to open the document.
+///
+/// Because we need to lazily initialize sub-resources (we don't know how deep
+/// the tree will be before reading the document), and that this is done in a
+/// context where no PDF reference allocator is available, `Resources` are
+/// originally created with the type parameter `R = ()`. The reference for each
+/// dictionnary will only be allocated in the next phase, once we know the shape
+/// of the tree, at which point `R` becomes `Ref`. No other value of `R` should
+/// ever exist.
+struct Resources<'a, R = Ref> {
+    /// The global reference to this resource dictionnary, or `()` if it has not
+    /// been allocated yet.
+    reference: R,
     /// Content of exported pages.
     pages: Vec<EncodedPage>,
+    /// The number of glyphs for all referenced languages in the document.
```

Comment:
I think it's a little confusing that the pages and languages are also considered `Resources` here. I would have expected `Resources` to only contain things for PDF-like resources.